src/membership/view.rs

```diff
@@ -1,15 +1,295 @@
+use crate::{
+    common::{ConfigId, Endpoint, NodeId},
+    error::{Error, Result},
+};
+use std::collections::HashSet;
+
+type Ring = crate::membership::ring::Ring<Endpoint>;
+
 #[derive(Debug, Clone)]
 pub struct View {
-    configuration: Configuration,
+    k: i32,
+    rings: Vec<Ring>,
+    seen: HashSet<NodeId>,
+    current_config: ConfigId,
+    should_update_configuration_id: bool,
+    current_configuration: Configuration,
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Default)]
 struct Configuration {
-    _p: (),
+    node_ids: Vec<NodeId>,
+    endpoints: Vec<Endpoint>,
 }
 
 impl View {
-    pub fn new() -> Self {
+    pub fn new(k: i32) -> Self {
+        assert!(k > 0);
+
+        let mut rings = Vec::with_capacity(k as usize);
+
+        for i in 0..k {
+            rings.push(Ring::new(i as u64));
+        }
+
+        Self {
+            k,
+            rings,
+            seen: HashSet::new(),
+            current_config: -1,
+            current_configuration: Configuration::default(),
+            should_update_configuration_id: true,
+        }
+    }
+
+    pub fn ring_add(&mut self, node: Endpoint, node_id: NodeId) -> Result<()> {
+        if self.is_node_present(&node_id) {
+            return Err(Error::new_uuid_already_seen());
+        }
+
+        if self.rings[0].contains(node.clone()) {
+            return Err(Error::new_node_already_in_ring());
+        }
+
+        for ring in &mut self.rings {
+            // Should this be append?
+            ring.insert(node.clone());
+        }
+
+        self.seen.insert(node_id);
+        self.should_update_configuration_id = true;
+
+        Ok(())
+    }
+
+    pub fn ring_delete(&mut self, node: &Endpoint) -> Result<()> {
+        if !self.rings[0].contains(node.clone()) {
+            return Err(Error::new_node_not_in_ring());
+        }
+
+        for ring in &mut self.rings {
+            ring.remove(node.clone());
+        }
+
+        self.should_update_configuration_id = true;
+
+        Ok(())
+    }
+
+    pub fn get_observers(&self, node: &Endpoint) -> Result<Vec<Endpoint>> {
+        if !self.rings[0].contains(node.clone()) {
+            return Err(Error::new_node_not_in_ring());
+        }
+
+        if self.rings[0].len() <= 1 {
+            return Ok(Vec::new());
+        }
+
+        let mut observers = Vec::new();
+
+        for ring in &self.rings {
```

Comment:
Again maybe a comment as to why this logic extracts the observers for a given node from the rings.